File: src/services/weatherService.js
import axios from 'axios';

const API_KEY = ''; // User should provide this
const BASE_URL = 'https://api.openweathermap.org/data/2.5';

export const fetchWeather = async (city = 'London') => {
  if (!API_KEY) {
    // Return mock data if no API key is provided
    return new Promise((resolve) => {
      setTimeout(() => {
        resolve({
          temp: 22,
          condition: 'Cloudy',
          icon: '04d',
          city: 'Mock City',
          humidity: 45,
          windSpeed: 12,
        });
      }, 1000);
    });
  }

  try {
    const response = await axios.get(`${BASE_URL}/weather`, {
      params: {
        q: city,
        appid: API_KEY,
        units: 'metric',
      },
    });

    return {
      temp: Math.round(response.data.main.temp),
      condition: response.data.weather[0].main,
      icon: response.data.weather[0].icon,
      city: response.data.name,
      humidity: response.data.main.humidity,
      windSpeed: response.data.wind.speed,
    };
  } catch (error) {
    console.error('Error fetching weather:', error);
    throw error;
  }
};

File: services/weather_service.py
import requests
import os
from datetime import datetime

class WeatherService:
    def __init__(self, api_key=None, city="London"):
        self.api_key = api_key or os.getenv("OPENWEATHER_API_KEY")
        self.city = city
        self.base_url = "http://api.openweathermap.org/data/2.5/weather"

    def get_weather(self):
        """Fetches weather data or returns mock data if API key is missing or request fails."""
        if not self.api_key or self.api_key == "YOUR_API_KEY":
            return self._get_mock_data()

        try:
            params = {
                "q": self.city,
                "appid": self.api_key,
                "units": "metric"
            }
            response = requests.get(self.base_url, params=params, timeout=5)
            response.raise_for_status()
            data = response.json()

            return {
                "temp": round(data["main"]["temp"]),
                "description": data["weather"][0]["description"].capitalize(),
                "icon": data["weather"][0]["icon"],
                "humidity": data["main"]["humidity"],
                "wind_speed": data["wind"]["speed"],
                "city": data["name"],
                "is_mock": False
            }
        except Exception as e:
            print(f"Weather API Error: {e}")
            return self._get_mock_data()

    def _get_mock_data(self):
        """Returns realistic mock weather data."""
        return {
            "temp": 22,
            "description": "Partly Cloudy",
            "icon": "02d",
            "humidity": 45,
            "wind_speed": 12,
            "city": "Mock City",
            "is_mock": True
        }

if __name__ == "__main__":
    service = WeatherService()
    print(service.get_weather())
